lib/widgets/lightbox.dart

```diff
@@ -113,6 +118,60 @@ class _CopyLinkButton extends StatelessWidget {
   }
 }
 
+class _DownloadImageButton extends StatelessWidget {
+  _DownloadImageButton({required this.url});
+
+  final downloader = ZulipBinding.instance.androidDownloadHost;
+  final Uri url;
+
+  @override
+  Widget build(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    final zulipLocalizations = ZulipLocalizations.of(context);
+    return IconButton(
+      tooltip: zulipLocalizations.lightboxDownloadImageTooltip,
+      icon: const Icon(Icons.download),
+      onPressed: () async {
+        final scaffoldMessenger = ScaffoldMessenger.of(context);
+        String message = zulipLocalizations.lightboxDownloadImageFailed;
+        try {
+          await requestStoragePermission();
+          Uri downloadUrl = url;
+          final internalLink = parseInternalLink(url, store);
+          if (internalLink is UserUploadLink){
+            if (!context.mounted) return;
+            final tempUrl = await ZulipAction.getFileTemporaryUrl(context, internalLink);
+            downloadUrl = tempUrl ?? url;
+          }
```

Comment:
If the link is a `UserUploadLink` and `ZulipAction.getFileTemporaryUrl` returns null, we don't expect it to be possible to download the file. In that case, `ZulipAction.getFileTemporaryUrl` will have already shown a dialog saying "Could not access uploaded file". We should stop there, without trying actions we don't expect to work (like `downloader.downloadFile`) and without showing additional UI feedback like a snackbar.